cassiopeia/solver/SharedMutationJoiningSolver.py

```diff
@@ -0,0 +1,298 @@
+"""
+This file stores the SharedMutationJoiningSolver. The inference procedure is 
+an agglomerative clustering procedure that joins samples that share the most
+identical character/state mutations.
+"""
+
+from typing import Callable, Dict, List, Optional, Tuple, Union
+
+import abc
+import networkx as nx
+import numba
+import numpy as np
+import pandas as pd
+
+from cassiopeia.data import CassiopeiaTree
+from cassiopeia.data import utilities as data_utilities
+from cassiopeia.solver import CassiopeiaSolver, solver_utilities
+
+
+class SharedMutationJoiningSolverError(Exception):
+    """An Exception class for SharedMutationJoiningSolver."""
+
+    pass
+
+
+class SharedMutationJoiningSolver(CassiopeiaSolver.CassiopeiaSolver):
+    """Shared-Mutation-Joining class for Cassiopeia.
+
+    Implements an iterative, bottom-up agglomerative clustering procedure. The
+    algorithm iteratively clusters the samples in the sample pool by the number
+    of shared mutations that they have in their character information. The
+    algorithm has theoretical guarantees on correctness given a sufficiently
+    large number of characters and bounds on edge lengths in the tree generative
+    process.
+
+    Args:
+        dissimilarity_function: Function that can be used to compute the
+            dissimilarity between samples.
+        prior_transformation: Function to use when transforming priors into
+            weights. Supports the following transformations:
+                "negative_log": Transforms each probability by the negative
+                    log (default)
+                "inverse": Transforms each probability p by taking 1/p
+                "square_root_inverse": Transforms each probability by the
+                    the square root of 1/p
+
+    Attributes:
+        dissimilarity_function: Function used to compute dissimilarity between
+            samples.
+        prior_transformation: Function to use when transforming priors into
+            weights.
+    """
+
+    def __init__(
+        self,
+        similarity_function: Callable[
+            [np.array, np.array, int, Dict[int, Dict[int, float]]], float
+        ],
+        prior_transformation: str = "negative_log",
+    ):
+
+        super().__init__(prior_transformation)
+
+        self.similarity_function = similarity_function
+
+    def solve(self, cassiopeia_tree: CassiopeiaTree) -> None:
+        """The solver routine for the SharedMutationJoiningSolver.
```

Comment:
Nit: `The solver...` -> `Solves a tree...`

I believe it's best code style to use verbs in the documentation titles.